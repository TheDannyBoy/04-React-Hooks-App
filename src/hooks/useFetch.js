import { useState, useEffect, useRef } from 'react'

export const useFetch = (url) => {
	const isMounted = useRef(true);
	const [state, setState] = useState({ data: null, loading: true, error: null });

	useEffect(() => {
		return () => {
			isMounted.current = false;
		};
	}, []);

	useEffect(() => {
		const initialState = { data: null, loading: true, error: null };
		setState(initialState);
		fetch(url)
			.then(resp => resp.json())
			.then(data => {
				if (isMounted.current) {
					setState({ loading: false, error: null, data });
				}
			})
			.catch(() => setState({ ...initialState, error: true }))

	}, [url]);

	return state;
}
