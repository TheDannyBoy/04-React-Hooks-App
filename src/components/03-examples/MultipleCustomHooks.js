import React from 'react';
import { useFetch } from '../../hooks/useFetch';
import { useCounter } from '../../hooks/useCounter';
import '../02-useEffect/effects.css';

export const MultipleCustomHooks = () => {
  const { counter, increment } = useCounter(1);
  const { data, loading } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);
  const { author, quote } = !!data && data[0];

  return (
    <div>
      <h1>Breaking Bad posts</h1>
      <hr/>

      { loading ?
        <div className="alert alert-info">
          Loading...
        </div>
        :
        <blockquote className='blockquote text-right'>
          <p className='mb-100'>{ quote }</p>
          <footer className='blockquote-footer'>{ author }</footer>
        </blockquote>
      }

      <button className='btn btn-primary' onClick={() => increment()}>Next Quote</button>
    </div>
  )
}
