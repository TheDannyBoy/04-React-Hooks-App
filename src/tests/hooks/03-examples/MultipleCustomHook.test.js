import { shallow } from 'enzyme';
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useCounter } from '../../../hooks/useCounter';
import { useFetch } from '../../../hooks/useFetch';

jest.mock('../../../hooks/useCounter');
jest.mock('../../../hooks/useFetch');

describe('Testing MultipleCustomHook hook', () => {

  beforeEach(() => {
    useCounter.mockReturnValue({ counter: 10, increment: () => {} });
  });

  test('should show correctly', () => {
    useFetch.mockReturnValue({ data: null, loading: true, error: null });
    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper).toMatchSnapshot();
  });

  test('should show data', () => {
    useFetch.mockReturnValue({
      data: [{ author: 'Dany', quote: '-' }],
      loading: false,
      error: null
    });
    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert').exists()).toBe(false);
  });
});
