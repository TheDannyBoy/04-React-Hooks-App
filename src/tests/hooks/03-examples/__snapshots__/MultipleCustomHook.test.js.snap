// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing MultipleCustomHook hook should show correctly 1`] = `
<div>
  <h1>
    Breaking Bad posts
  </h1>
  <hr />
  <div
    className="alert alert-info"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next Quote
  </button>
</div>
`;

exports[`Testing MultipleCustomHook hook should show data 1`] = `
<div>
  <h1>
    Breaking Bad posts
  </h1>
  <hr />
  <blockquote
    className="blockquote text-right"
  >
    <p
      className="mb-100"
    >
      -
    </p>
    <footer
      className="blockquote-footer"
    >
      Dany
    </footer>
  </blockquote>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next Quote
  </button>
</div>
`;
